asynchronous accept, synchronous secio

Stebalien opened the pull request:
Port of async connection accepting from #9

Stebalien committed:
accept connections asynchronously/update secio

Port of async connection accepting from #9

We need this to:

1. Perform the secio handshake immediately
2. Use a context as the dial timeout (to do this, we need to do the secio
handshake immediately).

Stebalien committed:
fix Accept test by actually completing crypto handshake.

Stebalien committed:
fix dial tests

One *can* write to a closed connection. It'll just fail eventually (but not
unless you send a bunch of data). The better test is to read from it.

Note: We only accept connections once they've been secured. This test *used* to
pass because we'd race and close the connection *before* the handshake
completed.

Stebalien committed:
downgrade non-error to info

Stebalien committed:
fix more incomplete dials

Stebalien committed:
tests: set accept timeout as well

whyrusleeping commented on dial.go:
```diff
@@ -19,6 +19,11 @@ import (
+// DialTimeout is the maximum duration a Dial is allowed to take.
+// This includes the time between dialing the raw network connection,
+// protocol selection as well the handshake, if applicable.
+var DialTimeout = 60 * time.Second
```
60 seconds now? Didnt this use to be 30 seconds?

Stebalien replied:
10 seconds, I think. However, this *also* includes secio negotiation (and now matches the accept timeout).

Honestly, they should probably *both* be 10.

Stebalien replied:
Changed back to 60. This is just a fallback timeout and should be conservative (users can always pass a context with a lower timeout).

Stebalien commented:
Note: these are currently failing `TestPNetIsUsed`.

Stebalien committed:
fix pnet test

No, you *may not* modify the input buffer to write.

Stebalien committed:
lower timeouts and remove an outdated comment

Stebalien committed:
gx update go-tcp-transport

Stebalien committed:
ci: update go version to 1.9

Stebalien committed:
make TestDialerCloseEarlySecure less racy

(and simplify it)

Before, one could finish the test and cancel the listener *before* it actually
accepted a connection.

This also vastly simplifies the test.

Kubuxu commented on dial.go:
```diff
@@ -19,6 +19,11 @@ import (
+// DialTimeout is the maximum duration a Dial is allowed to take.
+// This includes the time between dialing the raw network connection,
+// protocol selection as well the handshake, if applicable.
+var DialTimeout = 30 * time.Second
```
This limits the maximum ping to quite a low value. To finish handshake we need about 5 round-trips which makes max RTT be 6s.

Stebalien replied:
6s RTT is *really* long. However, I could make these finer grained (I'll go do that, actually).

Stebalien replied:
Ok, I just went ahead and bumped this to 60s (what we had before for the secio handshake). That should be sufficient, IMO.

I did not add fine-grained timeouts. I'd like to respect the context timeout as much as possible. We can mess with that later but I'd like to merge this ASAP (I've already let this whole mess go *way* too long).

Stebalien committed:
Bump the dial/accept timeouts to 60s

This matches the previous secio timeout and should, conservatively, be enough.
We can optimize/rethink this later but not in this PR.

Stebalien committed:
gx publish 1.7.0
